### PR TITLE
[Model] Optimize Sarvam MLA routing and preserve FP32 router logits

### PR DESCRIPTION
**Pull request by harshit-sarvam**
## Purpose

Fix Sarvam MLA's missing flat-routing configuration and preserve the configured router-logit dtype through expert selection. On the tested BF16 B200 setup, these changes enable the supported TRTLLM MoE path and reduce router overhead while removing an unnecessary BF16 narrowing of FP32 logits.

The source diff is confined to `vllm/model_executor/models/sarvam.py`:

- Default absent or `None` `n_group` and `topk_group` to `1`. The model uses flat top-8-of-128 routing (verified). A single group expresses that configuration and allows the supported TRTLLM implementation to be selected. Explicit checkpoint values are preserved.
- Replace `nn.Linear` plus input/output casts with the existing `GateLinear(out_dtype=router_dtype)`. With the default FP32 router output, the tested path uses BF16 inputs/weights and FP32 accumulation/output, and keeps those logits in FP32 for routing. Previously, the FP32 linear result was narrowed to the hidden-state dtype before expert selection. Other configured router output dtypes remain respected.

### Duplicate-work check

Searched open upstream PRs for `sarvam`, `sarvam routing`, and `sarvam GateLinear`; no open PR addressing these router changes was found.

## Test Plan

Validation below combines local checks on the merged branch with historical GPU experiments on the constituent changes. **The GPU results predate the latest upstream merge; they are not a fresh evaluation of this merged head.** This is a draft for human review and post-merge validation.

### Performance

Plain Sarvam step-1350 checkpoint, BF16, 2× NVIDIA B200, TP=2, no speculative model. Gate comparisons used EP off, compilation enabled, prefix caching disabled, maximum model length 32,768, and fixed real-data prompts truncated to 8,192 tokens. Each case used five timing repetitions and medians. Decode step time was estimated as `(median time for 136 output tokens − median time for 8 output tokens) / 128`, with EOS ignored. Prefill throughput uses the corresponding timing intercept and includes admission/scheduler overhead.

| Batch | Control decode ms/step | Repeated control | GateLinear decode ms/step | Speedup against both controls |
| --- | ---: | ---: | ---: | ---: |
| 1 | 4.844 | 4.664 | 4.473 | 1.043–1.083× |
| 32 | 15.806 | 15.747 | 15.304 | 1.029–1.033× |

| Batch | Control prefill tokens/s | Repeated control | GateLinear prefill tokens/s | Speedup against both controls |
| --- | ---: | ---: | ---: | ---: |
| 1 | 62,419 | 62,298 | 67,078 | 1.075–1.077× |
| 32 | 67,655 | 67,035 | 70,294 | 1.039–1.049× |

Separately, the flat-routing defaults switched the tested MoE path from CUTLASS to TRTLLM and measured **1.11–1.17× decode throughput and 1.03–1.04× prefill throughput**.

### Routing precision verification

An isolated router experiment used real gate weights from all 31 MoE layers, synthetic layer-normalization-scaled inputs, 65,536 inputs per layer, and three seeds on B200: **6,094,848 layer/input comparisons**.

| Candidate compared with the old FP32-gate → BF16-logit path | Inputs with a different selected expert set |
| --- | ---: |
| BF16 gate, still narrow logits to BF16 | 630 / 6,094,848 (0.0103%) |
| BF16 gate, retain FP32 logits | 133,833 / 6,094,848 (2.1958%) |
| FP32 gate, retain FP32 logits | 133,842 / 6,094,848 (2.1960%) |

Removing the logit downcast accounts for the dominant change in expert selection. This is intentionally not bit-exact to the old path: rounding logits before sigmoid/bias/top-k can alter close routing decisions. These counts measure expert-set disagreement, not task error rates, and do not establish exact equivalence between the two FP32-output candidates or full training/inference numerical parity.
Separate evals like AIME, GPQA, TAU2, etc were conducted to ensure parity in quality.

### End-to-end repeatability

Each run scored the same **687,500 positions across 200 fixed prompts**, up to 4,096 tokens each, at **concurrency 1**, with 2×B200, TP=2, EP enabled, BF16, compilation enabled, and prefix caching disabled. A disagreement means a different top-1 next token given the same dataset prefix; this is teacher-forced agreement, not diverging free-running generations. Captures contain top-1 token IDs and saved log probabilities, not full logit tensors.

| Historical build | Selected MoE backend | Top-1 differences against its repeat |
| --- | --- | ---: |
| Base commit (main) | CUTLASS | 67,408 / 687,500 (9.805%) |
| Flat-routing defaults (`879e62199d`) | TRTLLM | 0; saved log probabilities bit-identical |
| GateLinear / FP32 logits (`647516fb80`) | TRTLLM | 0; saved log probabilities bit-identical |

Each repeat pair shared an allocation; different historical builds ran on different nodes using the dependencies installed at experiment time. The measured repeatability boundary is the flat-routing/backend-selection change. This does not isolate the internal cause of the earlier variance or establish determinism at higher concurrency.

Cross-implementation diagnostics also showed that scale placement and bias handling affect Sarvam-versus-DeepSeek agreement. Those diagnostic modifications are excluded from this PR; DeepSeek agreement is not treated as proof of task accuracy.

## Test Result

Checks run on the merged branch:

```text
uv tool run ruff check vllm/model_executor/models/sarvam.py        PASS
uv tool run ruff format --check vllm/model_executor/models/sarvam.py  PASS
git diff --check upstream/main...HEAD                           PASS
```

AI assistance was used to prepare and review this contribution and summarize the recorded experiments. Human review and final validation remain required.

### CODE REVIEW

**Comment by harshit-sarvam** [comment]
@gau-nernst can you please help get this merged!

**Comment by gau-nernst** [comment]
@harshit-sarvam Can you fix DCO? Thanks

**Comment by harshit-sarvam** [comment]
@gau-nernst fixed DCO

**Comment by harshit-sarvam** [comment]
Several jobs failed during checkout due to Git mirror lock timeouts.
AMD Basic Correctness also segfaulted in test_mem.py.

**Comment by harshit-sarvam** [comment]
@gau-nernst can you please run retry?

**Review by ZJY0516 — APPROVED** [review]
